robot/impl/local_robot.go

```diff
@@ -709,12 +723,135 @@ func (r *localRobot) updateWeakDependents(ctx context.Context) {
-func (r *localRobot) FrameSystemConfig(
-	ctx context.Context,
-	additionalTransforms []*referenceframe.LinkInFrame,
-) (framesystemparts.Parts, error) {
-	return r.frameSvc.Config(ctx, additionalTransforms)
+// Config returns the info of each individual part that makes up the frame system
+// The output of this function is to be sent over GRPC to the client, so the client
+// can build its frame system. requests the remote components from the remote's frame system service.
+// NOTE(RDK-258): If remotes can trigger a local robot to reconfigure, you don't need to update remotes in every call.
+func (r *localRobot) FrameSystemConfig(ctx context.Context) (*framesystem.Config, error) {
+	r.fsMu.RLock()
+	defer r.fsMu.RUnlock()
+
+	localParts, err := r.getLocalParts(ctx)
+	if err != nil {
+		return nil, err
+	}
+	remoteParts, err := r.getRemoteParts(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &framesystem.Config{Parts: append(localParts, remoteParts...)}, nil
+}
+
+// getLocalParts collects and returns the physical parts of the robot that may have frame info,
+// excluding remote robots and services, etc from the robot's config.Config.
+func (r *localRobot) getLocalParts(ctx context.Context) ([]*referenceframe.FrameSystemPart, error) {
+	cfg, err := r.Config(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	parts := make([]*referenceframe.FrameSystemPart, 0)
+	for _, component := range cfg.Components {
+		if component.Frame == nil { // no Frame means dont include in frame system.
+			continue
+		}
+
+		if component.Name == referenceframe.World {
+			return nil, errors.Errorf("cannot give frame system part the name %s", referenceframe.World)
+		}
+		if component.Frame.Parent == "" {
+			return nil, errors.Errorf("parent field in frame config for part %q is empty", component.Name)
+		}
+		cfgCopy := &referenceframe.LinkConfig{
+			ID:          component.Frame.ID,
+			Translation: component.Frame.Translation,
+			Orientation: component.Frame.Orientation,
+			Geometry:    component.Frame.Geometry,
+			Parent:      component.Frame.Parent,
+		}
+		if cfgCopy.ID == "" {
+			cfgCopy.ID = component.Name
+		}
+		model, err := r.extractModelFrameJSON(component.ResourceName())
+		if err != nil && !errors.Is(err, referenceframe.ErrNoModelInformation) {
+			// When we have non-nil errors here, it is because the resource is not yet available.
+			// In this case, we will exclude it from the FS.
+			// When it becomes available, it will be included.
+			continue
+		}
+		lif, err := cfgCopy.ParseConfig()
+		if err != nil {
+			return nil, err
+		}
+
+		parts = append(parts, &referenceframe.FrameSystemPart{FrameConfig: lif, ModelFrame: model})
+	}
+	return parts, nil
+}
+
+func (r *localRobot) getRemoteParts(ctx context.Context) ([]*referenceframe.FrameSystemPart, error) {
+	cfg, err := r.Config(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	remoteParts := make([]*referenceframe.FrameSystemPart, 0)
+	for _, remoteName := range r.RemoteNames() {
```

Comment:
Why not loop over `conf.Remotes` instead of `r.RemoteNames()`? Should avoid the necessity for `getRemoteRobotConfig()`.

Reply:
@biotinker, under the hood this function is just iterating over `conf.Remotes` so I think this could be a useful simplification

Reply:
> under the hood this function is just iterating over conf.Remotes

I don't think that's true. It calls the resource manager's [`RemoteNames`](https://github.com/viamrobotics/rdk/blob/main/robot/impl/resource_manager.go#L261) which looks through the resource graph for all nodes with the remote API. You were, however, looping through `conf.Remotes` in `getRemoteRobotConfig` to look for name matches, which means remotes from `RemoteNames` that were not in `conf.Remotes` were being ignored (not sure if this is what you meant @raybjork). LGTM now.